internal/pkg/stream/stream.go

```diff
@@ -13,69 +13,61 @@ import (
 	streamerMaxFetchIntervalDurationMs = 32000 // The maximum duration that a client should wait until Fetch is called again.
+	streamerMinFetchIntervalDurationMs = 1000  // The minimum duration that a client should wait until Fetch is called again.
 )
 
 // Streamer is the interface that groups methods to periodically retrieve events,
 // publish them to subscribers, and stop publishing once there are no more events left.
 type Streamer interface {
 	// Fetch fetches events, updates the internal state of the Streamer with new events and returns the next time
 	// the Fetch call should be attempted. On failure, Fetch returns an error.
-	Fetch() (next time.Time, err error)
+	Fetch() (next time.Time, done bool, err error)
 
 	// Notify publishes all new event updates to subscribers.
 	Notify()
 
 	// Close notifies all subscribers that no more events will be sent.
 	Close()
-
-	// Done returns a channel that's closed when there are no more events to Fetch.
-	Done() <-chan struct{}
 }
 
 // Stream streams event updates by calling Fetch followed with Notify until there are no more events left.
 // If the context is canceled or Fetch errors, then Stream short-circuits and returns the error.
 func Stream(ctx context.Context, streamer Streamer) error {
 	defer streamer.Close()
-
-	var next time.Time
-	var err error
+	var fetchDelay time.Duration // By default there is no delay.
 	for {
-		var fetchDelay time.Duration // By default there is no delay.
-		if now := time.Now(); next.After(now) {
-			fetchDelay = next.Sub(now)
-		}
-
 		select {
 		case <-ctx.Done():
 			return ctx.Err()
-		case <-streamer.Done():
-			// No more events to Fetch, flush and exit successfully.
-			streamer.Notify()
-			return nil
 		case <-time.After(fetchDelay):
-			next, err = streamer.Fetch()
+			next, done, err := streamer.Fetch()
 			if err != nil {
 				return err
 			}
 			streamer.Notify()
+			if done {
+				return nil
+			}
+			fetchDelay = time.Until(next)
 		}
 	}
 }
 
 // nextFetchDate returns a time to wait using random jitter and exponential backoff.
 func nextFetchDate(clock clock, rand func(int) int, retries int) time.Time {
-	// waitMs := rand.Intn( 							// Get a random integer between 0 and ...
+	// waitMs := rand.Intn( 							// Get a random integer between streamerMinFetchIntervalDurationMs and ...
 	// 	min( 											// the minimum of ...
 	// 		streamerMaxFetchIntervalDuration,           // the max fetch interval and ...
 	// 		streamerFetchIntervalDuration*(1<<retries), // d*2^r, where r=retries and d= the normal
-	// 	),
-	// )
-	waitMs := rand(
-		min(
-			streamerMaxFetchIntervalDurationMs,
-			streamerFetchIntervalDurationMs*(1<<retries),
-		),
-	)
+	// 	)-streamerMinFetchIntervalDurationMs
+	// ) + streamerMinFetchIntervalDurationMs
+	waitMs :=
+		rand(
+			min(
+				streamerMaxFetchIntervalDurationMs,
+				streamerFetchIntervalDurationMs*(1<<retries),
+			)-streamerMinFetchIntervalDurationMs,
```

Comment:
I am having difficulty parsing this 😅 even after reading the comments above from Danny.
After plugging in the numbers it makes sense! I think this might be equivalent too:
```
jitter := rand(
	min(
		streamerMaxFetchIntervalDurationMs - streamerMinFetchIntervalDurationMs,
		streamerFetchIntervalDurationMs*(1<<retries),
	)
)
waitMs := streamerMinFetchIntervalDurationMs + jitter
```

Reply:
hmm it's not equivalent when `retries` is a small number like 0. In this case the current implementation has 
```
[streamerMinFetchIntervalDurationMs, streamerFetchIntervalDurationMs*(1<<retries))
```
where as yours is
```
[streamerMinFetchIntervalDurationMs, streamerFetchIntervalDurationMs*(1<<retries)+streamerMinFetchIntervalDurationMs)
```

I'll add a URL with more info about this https://www.educative.io/answers/how-to-generate-random-numbers-in-a-given-range-in-go